mtt/src/mtt_types.rs

```diff
@@ -118,3 +116,226 @@ impl MttL1EntryType {
+        let range_addr = gb_range_start.checked_add_pages(i * 16 * 1024).unwrap();
+        let mtt_entry = get_l2_entry(range_addr);
+        match mtt_entry {
+            MttNonConfidential1G if mtt_entry == first_entry_type => {
+                set_l2_entry(range_addr, MttL2EntryType::non_confidential_64_mb_range())
+            }
+            MttConfidential1G if mtt_entry == first_entry_type => {
+                set_l2_entry(range_addr, MttL2EntryType::confidential_64_mb_range())
+            }
+            _ => {
+                panic!(
+                    "fracture_1gb_range() error: start: {:x}, current: {:x} first_entry: {first_entry_type:?} current_entry: {mtt_entry:?}",
+                    gb_range_start.bits(), range_addr.bits()
+                )
+            }
+        }
+    }
+}
+
+// Changes the L2 type of 64MB range containing the address to 4Kx16KB.
+// All pages in the range will retain the original MTT type (non-confidential or confidential).
+fn fracture_64mb_range(phys_addr: SupervisorPageAddr) {
+    let mtt_entry = get_l2_entry(phys_addr);
+    let mtt_l1_page = get_l1_page(phys_addr);
+    let l1_page_addr = mtt_l1_page.bits();
+    match mtt_entry {
+        Mtt2MBPages(value) => {
+            // Mark the entire 64MB as non-confidential at the outset
+            // Safety: mtt_l1_page is a supervisor page address that's owned by us, and is at least 4K in size
+            unsafe {
+                core::ptr::write_bytes(l1_page_addr as *mut u8, 0, PageSize::Size4k as usize);
+            }
+            for i in 0u64..=31 {
+                // 01010101b encoding for 4-pages of confidential memory
+                const L1_MTT_CONFIDENTIAL_16K_ENCODING: u8 = 0x55;
+                // Mark the 2MB sub-range as confidential if applicable (128x16K)
+                if value & (1 << i) == MttMemoryType::Confidential as u64 {
+                    // Safety: mtt_l1_page is a supervisor page address that's owned by us, and is at least 4K in size
+                    // The maximum index is 31*128 from the base and we write 128-bytes at a time
+                    unsafe {
+                        core::ptr::write_bytes(
+                            (l1_page_addr + (i * 128)) as *mut u8,
+                            L1_MTT_CONFIDENTIAL_16K_ENCODING,
+                            128,
+                        );
+                    }
+                }
+            }
+
+            set_l2_entry(phys_addr, MttL2EntryType::mtt_l1_dir(mtt_l1_page));
+        }
+        _ => panic!(
+            "fracture_64mb_range() error: address: {:x}, entry: {mtt_entry:?}",
+            phys_addr.bits()
+        ),
+    }
+}
+
+/// Sets the type of the 4K page corresponding to `phys_addr` to the type specified
+/// in `memory_type`
+pub fn set_4k_page_type(phys_addr: SupervisorPageAddr, memory_type: MttMemoryType) {
+    match get_l2_entry(phys_addr) {
```

Comment:
Did you try to define a type for an l2_entry? The pattern of repeatedly calling `get_entry` while passing around the address feels like it could be greatly simplified if the entry could be passed directly. It'll also help to minimize error handling propagation that is kind of scattered in this current implementation.

Reply:
Ack.

Reply:
See if the improved error handling has address the above. Were you thinking of encapsulating the l2_entry data (including the mtt_index and so forth) into a type that be passed around?

Reply:
The error handling looks a lot better, thanks!

For the entry type, I'm guessing that if you build a wrapper type as above to interfaced with the MTT pages, returning a mut ref to the entries will be more natural than calling get and set sequentially with an address. Happy to chat about this on VC tomorrow too if you want.

Reply:
A VC chat sounds good...see if the latest iteration makes more sense.